# The Series Data Structure

To open this file as a Google Colab, please refer to the direct link in the ReadMe section (2nd point of the Miscellaneous Chapter). The series is one of the core data structures in pandas. You think of it a cross between a list and a dictionary. The items are all stored in an order and there's labels with which you can retrieve them. 

In [ ]:
import pandas as pd


Jupyter Notebook provides an extended help and show the documentation, using the question mark operator `?`

In [ ]:
pd.Series?

In [ ]:
animals = ['Tiger', 'Bear', 'Moose']
pd.Series(animals)

0    Tiger
1     Bear
2    Moose
dtype: object

In [ ]:
numbers = [1, 2, 3]
pd.Series(numbers)

0    1
1    2
2    3
dtype: int64

Underneath panda stores series values in a typed array using the Numpy library. This offers significant speed-up when processing data versus traditional python lists.

There's some other typing details that exist for performance that are important to know. The most important is how Numpy and thus pandas handle missing data. In Python, we have the `None` type to indicate a lack of data. But what do we do if we want to have a typed list like we do in the series object?

In [ ]:
animals = ['Tiger', 'Bear', None]
pd.Series(animals)

0    Tiger
1     Bear
2     None
dtype: object

 If we create a list of numbers, integers or floats, and put in the None type, Pandas automatically converts this to a special floating point value designated as NAN, which stands for not a number.

In [ ]:
numbers = [1, 2, None]
pd.Series(numbers)

0    1.0
1    2.0
2    NaN
dtype: float64

NAN is not none and when we try the equality test, it's false.

In [ ]:
import numpy as np
np.nan == None

False

It turns out that you actually can't do an equality test of NAN to itself. When you do, the answer is always false.

In [ ]:
np.nan == np.nan


False

 You need to use special functions to test for the presence of not a number, such as the Numpy library `isnan()`.

In [ ]:
np.isnan(np.nan)

True

A series can be created from dictionary data. 

In [ ]:
sports = {'Football': 'Germany',
          'Golf': 'Scotland',
          'Sumo': 'Japan',
          'Taekwondo': 'South Korea'}
s = pd.Series(sports)
s

Football         Germany
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

In [ ]:
s.index

Index(['Football', 'Golf', 'Sumo', 'Taekwondo'], dtype='object')

You could also separate your index creation from the data by passing in the index as a list explicitly to the series.

In [ ]:
s = pd.Series(['Tiger', 'Bear', 'Moose'], index=['India', 'America', 'Canada'])
s

India      Tiger
America     Bear
Canada     Moose
dtype: object

In the following example, we pass in a dictionary of four items but only two are preserved in the series object because of the index list. We see that "Hockey" has been added but since it's also in the index list. But it has no value associated with it.

In [ ]:
sports = {'Football': 'Germany',
          'Golf': 'Scotland',
          'Sumo': 'Japan',
          'Taekwondo': 'South Korea'}
s = pd.Series(sports, index=['Golf', 'Sumo', 'Hockey'])
s

Golf      Scotland
Sumo         Japan
Hockey         NaN
dtype: object

# Querying a Series

A Pandas Series can be queried, either by the index position or the index label. As we saw, if you don't give an index to the series, the position and the label are effectively the same values. To query by numeric location, starting at zero, use the `iloc` attribute. To query by the index label, you can use the `loc` attribute. 

In [ ]:
import pandas as pd

In [ ]:
sports = {'Football': 'Germany',
          'Golf': 'Scotland',
          'Sumo': 'Japan',
          'Taekwondo': 'South Korea'}
s = pd.Series(sports)
s

Football         Germany
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

Access a row by integer index

In [ ]:
s.iloc[3]

'South Korea'

Access a row by label

In [ ]:
s.loc['Golf']

'Scotland'

Keep in mind that `iloc` and `loc` are not methods, they are attributes. So you don't use parentheses to query them, but square brackets instead, which we'll call the indexing operator. Though in Python, this calls get and set an item methods depending on the context of its use.

if you pass in an integer parameter, the operator will behave as if you want it to query via the `iloc` attribute.

In [ ]:
s[3]

'South Korea'

If you pass in an object, it will query as if you wanted to use the label based `loc` attribute.

In [ ]:
s['Golf']

'Scotland'

Here's an example using some new sports data, where countries are indexed by integer.

In [ ]:
sports = {99: 'Germany',
          100: 'Scotland',
          101: 'Japan',
          102: 'South Korea'}
s = pd.Series(sports)
s

99         Germany
100       Scotland
101          Japan
102    South Korea
dtype: object

If we try and call `s[0]`, we get a key error

In [ ]:
s[0] # we get a key error, because there's no item in the sports list with an index of zero. 

KeyError: 0

In [ ]:
s.iloc[0] # Instead we have to call iloc explicitly if we want the first item.

'Germany'

In [ ]:
s[99] # the first entry has the index of 99

'Germany'

Let's talk about working with the data.

A data Series of floating point values.

In [ ]:
s = pd.Series([100.00, 120.00, 101.00, 3.00])
s

0    100.0
1    120.0
2    101.0
3      3.0
dtype: float64

An iteration over all of the items in the series and adds them together to get a total.

In [ ]:
total = 0
for item in s:
    total+=item
print(total)

324.0


This works, but it's slow. 

Faster approach using the NumPy sum method.

In [ ]:
import numpy as np

total = np.sum(s)
print(total)

324.0


In [ ]:
#this creates a big series of random numbers
s = pd.Series(np.random.randint(0,1000,10000)) #low=0, high=1000, size=10000
s.head()

0    972
1    404
2    964
3    809
4    350
dtype: int64

In [ ]:
len(s)

10000

A Jupyter Notebook magic function begin with a percentage sign. If we type this sign and then hit the Tab key, we can see a list of the available magic functions. 

In [ ]:
%

UsageError: Line magic function `%` not found.


The function we're going to use is called `timeit`; this function will run our code a few times to determine, on average, how long it takes.

In [ ]:
%%timeit -n 100
summary = 0
for item in s:
    summary+=item

1.62 ms ± 278 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Now let's try with `sum()` function (with vectorization)

In [ ]:
%%timeit -n 100
summary = np.sum(s)

139 µs ± 20.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


this is much faster!

A related feature in Pandas and NumPy is called **broadcasting**. With broadcasting, you can apply an operation to every value in the series, changing the series.

In [ ]:
print(s.head())
s+=2 #adds two to each item in s using broadcasting
print(s.head())

0    972
1    404
2    964
3    809
4    350
dtype: int64
0    974
1    406
2    966
3    811
4    352
dtype: int64


In [ ]:
for label, value in s.iteritems():
    # s.set_value(label, value+2) # set_value is deprecated
    s.at[label]=value+2
s.head()

0    976
1    408
2    968
3    813
4    354
dtype: int64

Let's try and time the two approaches for increasing the value

In [ ]:
%%timeit -n 10
s = pd.Series(np.random.randint(0,1000,10000))
for label, value in s.iteritems():
    s.loc[label]= value+2

832 ms ± 118 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [ ]:
%%timeit -n 10
s = pd.Series(np.random.randint(0,1000,10000))
s+=2


564 µs ± 160 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


Amazing. Not only is it significantly faster, but it's more concise and maybe even easier to read too.

One last note on using the indexing operators to access series data. The `.loc` attribute lets you not only modify data in place, but also add new data as well.

Pandas will automatically change the underlying NumPy types as appropriate.

In [ ]:
s = pd.Series([1, 2, 3])
print(s)
s.loc['Animal'] = 'Dog'
print(s)

0    1
1    2
2    3
dtype: int64
0           1
1           2
2           3
Animal    Dog
dtype: object


We end this chapter by showing example where index values are not unique, and this makes a data Series different conceptually, that a relational database might be

Revisiting the issue of countries and their national sports, it turns out that many countries seem to like this game cricket. We go back to our original series on sports. It's possible to create a new series object with multiple entries for cricket, and then use append to bring these together. There are a couple of important considerations when using append. First, Pandas is going to take your series and try to infer the best data types to use. In this example, everything is a string, so there's no problems here.

Second, the append method doesn't actually change the underlying series. It instead returns a new series which is made up of the two appended together. 

In [ ]:
original_sports = pd.Series({'Football': 'Germany',
                             'Golf': 'Scotland',
                             'Sumo': 'Japan',
                             'Taekwondo': 'South Korea'})

cricket_loving_countries = pd.Series(['Australia',
                                      'Barbados',
                                      'Pakistan',
                                      'England'], 
                                   index=['Cricket',
                                          'Cricket',
                                          'Cricket',
                                          'Cricket'])

all_countries = original_sports.append(cricket_loving_countries)

Printing the original series of values and seeing that they haven't changed.

In [ ]:
original_sports

Football         Germany
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

This is actually a significant issue for new Pandas users who are used to objects being changed in place. So watch out for it, not just with append but with other Pandas functions as well.

In [ ]:
cricket_loving_countries

Cricket    Australia
Cricket     Barbados
Cricket     Pakistan
Cricket      England
dtype: object

In [ ]:
all_countries

Football         Germany
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
Cricket        Australia
Cricket         Barbados
Cricket         Pakistan
Cricket          England
dtype: object

Finally, we see that when we query the appended series for those who have cricket as their national sport, we don't get a single value, but a series itself.

In [ ]:
all_countries.loc['Cricket']

Cricket    Australia
Cricket     Barbados
Cricket     Pakistan
Cricket      England
dtype: object

# The DataFrame Data Structure

The DataFrame data structure is the heart of the Panda's library. It's a primary object that you'll be working with in data analysis and cleaning tasks.

The DataFrame is conceptually a two-dimensional series object, where there's an index and multiple columns of content, with each column having a label. In fact, the distinction between a column and a row is really only a conceptual distinction. And you can think of the DataFrame itself as simply a two-axes labeled array.

## Data Frame creation
You can create a DataFrame in many different ways, some of which you might expect. For instance, you can use a group of series, where each series represents a row of data. Or you could use a group of dictionaries, where each dictionary represents a row of data.

We are going to create three purchase order records as series objects for a sort of fictional store.

In [ ]:
import pandas as pd
purchase_1 = pd.Series({'Name': 'Matthias',
                        'Item Purchased': 'Dog Food',
                        'Cost': 22.50})
purchase_2 = pd.Series({'Name': 'Thomas',
                        'Item Purchased': 'Kitty Litter',
                        'Cost': 2.50})
purchase_3 = pd.Series({'Name': 'Christina',
                        'Item Purchased': 'Bird Seed',
                        'Cost': 5.00})

df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])
df

Name Item Purchased  Cost
Store 1   Matthias       Dog Food  22.5
Store 1     Thomas   Kitty Litter   2.5
Store 2  Christina      Bird Seed   5.0

In this example, we wanted to select data associated with Store 2, so we query the loc attribute with one parameter.

## Data Access
Similar to the series, we can extract data using the `iloc` and `loc` attributes.



In [ ]:
df.iloc[0]

Name              Matthias
Item Purchased    Dog Food
Cost                  22.5
Name: Store 1, dtype: object

Because the DataFrame is two-dimensional, passing a single value to the `loc` indexing operator will return a series if there's only one row to return.

In the following example, we want to select data associated with Store 2, so we query the `loc` attribute with one parameter.

In [ ]:
df.loc['Store 2']

Name              Christina
Item Purchased    Bird Seed
Cost                      5
Name: Store 2, dtype: object

 We can check the data type of the return using the python type function.

In [ ]:
type(df.loc['Store 2'])

pandas.core.series.Series

It's important to remember that the indices and column names along either axes, horizontal or vertical, could be non-unique.

For instance, in this example, we see two purchase records for Store 1 as different rows.

In [ ]:
df.loc['Store 1']

Name Item Purchased  Cost
Store 1  Matthias       Dog Food  22.5
Store 1    Thomas   Kitty Litter   2.5

We see that Matthias and Thomas both shop at the same pets supply store.

In this case, the return value is a Data Frame

In [ ]:
type(df.loc['Store 1'])

pandas.core.frame.DataFrame

One of the powers of the Panda's DataFrame is that you can quickly select data based on multiple axes.

For instance, if you wanted to just list the costs for Store 1, you would supply two parameters to `.loc`, one being the row **index** and the other being the **column** name.

In [ ]:
df.loc['Store 1', 'Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

## Column Selection

What if we just wanted to do column selection and just get a list of all of the costs? Well, there's a couple of options.

### 1. Transpose Data Frame
You can get a transpose of the DataFrame, using the capital T attribute, which swaps all of the columns and rows.

In [ ]:
df.T

Store 1       Store 1    Store 2
Name            Matthias        Thomas  Christina
Item Purchased  Dog Food  Kitty Litter  Bird Seed
Cost                22.5           2.5          5

And we can then use the .loc method

In [ ]:
df.T.loc['Cost']

Store 1    22.5
Store 1     2.5
Store 2       5
Name: Cost, dtype: object

This works, but it's **pretty ugly**.

### 2. Indexing Operator

In a Pandas DataFrame, columns always have a name;  so this selection is always label based. 

(For those familiar with relational databases, this operator is analogous to column projection.)

In [ ]:
df['Cost']

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: float64

### 3. Operation Chaining
Finally, since the result of using the indexing operators, the DataFrame or series, you can chain operations together.

In [ ]:
# loc operation then indexing operation
s1 = df.loc['Store 1']
s1['Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

In [ ]:
# chaining of loc and indexing
df.loc['Store 1']['Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

<div class="alert alert-block alert-warning">
But chaining can come with some costs and is best avoided if you can use another approach. In particular, chaining tends to cause Pandas to return a copy of the DataFrame instead of a view on the DataFrame. For selecting a data, this is not a big deal, though it might be slower than necessary. If you are changing data though, this is an important distinction and can be a source of error.
</div>

### 4. Full slice with Loc

As we saw, .loc does row selection, and it can take two parameters, the row index and the list of column names. italso supports slicing. If we wanted to select all rows, we can use a column to indicate a full slice from beginning to end. And then add the column name as the second parameter as a string. In fact, if we wanted to include multiply columns, we could do so in a list. And Pandas will bring back only the columns we have asked for.

In [ ]:
df.loc[:,['Name', 'Cost']]


Name  Cost
Store 1   Matthias  22.5
Store 1     Thomas   2.5
Store 2  Christina   5.0

The key concepts to remember are that the rows and columns are really just for our benefit. Underneath this is a two axes labeled array. 

## Dropping Data

It's easy to delete data in series and DataFrames, and we can use the drop function to do so. This function takes a single parameter, which is the index or roll label, to drop.

In [ ]:
df.drop('Store 1')

Name Item Purchased  Cost
Store 2  Christina      Bird Seed   5.0

The drop function doesn't change the DataFrame by default. And instead, returns to you a copy of the DataFrame with the given rows removed. 

We can see that our original DataFrame is still intact.

In [ ]:
df

Name Item Purchased  Cost
Store 1   Matthias       Dog Food  22.5
Store 1     Thomas   Kitty Litter   2.5
Store 2  Christina      Bird Seed   5.0

Let's make a copy with the copy method and do a drop on it instead.

In [ ]:
copy_df = df.copy()
copy_df = copy_df.drop('Store 1')
copy_df

Name Item Purchased  Cost
Store 2  Christina      Bird Seed   5.0

This is a very typical pattern in Pandas, where in place changes to a DataFrame are only done if need be, usually on changes involving indices.

**So it's important to be aware of.**

In [ ]:
copy_df.drop?

 Drop has two interesting optional parameters.

- The first is called in place, and if it's set to true, the DataFrame will be updated in place, instead of a copy being returned.

- The second parameter is the axes, which should be dropped.

    - By default, this value is 0, indicating the row axes.

    - But you could change it to 1 if you want to drop a column.


There is a second way to drop a column, however. This is directly through the use of the indexing operator, using the `del` keyword.


In [ ]:
del copy_df['Name']
copy_df

Item Purchased  Cost
Store 2      Bird Seed   5.0

This way of dropping data, however, takes immediate effect on the DataFrame and does not return a view.

## Adding a new column

Finally, adding a new column to the DataFrame is as easy as assigning it to some value.

For instance, if we wanted to add a new location as a column with default value of none, we could do so by using the assignment operator after the square brackets.

In [ ]:
df['Location'] = None
df

Name Item Purchased  Cost Location
Store 1   Matthias       Dog Food  22.5     None
Store 1     Thomas   Kitty Litter   2.5     None
Store 2  Christina      Bird Seed   5.0     None

This broadcasts the default value to the new column immediately.


**In conclusion**, we presented in this chapter: DataFrame creation, data access, column selection, dropping of data and adding a new column. 

# Dataframe Indexing and Loading

The common work flow is to read your data into a DataFrame then reduce this DataFrame to the particular columns or rows that you're interested in working with. As you've seen, the Pandas toolkit tries to give you views on a DataFrame. This is much faster than copying data and much more memory efficient too.

But it does mean that if you're manipulating the data you have to be aware that any changes to the DataFrame you're working on may have an impact on the base data frame you used originally.

Here's an example using our same purchasing DataFrame from earlier. We can create a series based on just the cost category using the square brackets.

## Data Index

In [ ]:
import pandas as pd
purchase_1 = pd.Series({'Name': 'Matthias',
                        'Item Purchased': 'Dog Food',
                        'Cost': 22.50})
purchase_2 = pd.Series({'Name': 'Thomas',
                        'Item Purchased': 'Kitty Litter',
                        'Cost': 2.50})
purchase_3 = pd.Series({'Name': 'Christina',
                        'Item Purchased': 'Bird Seed',
                        'Cost': 5.00})

df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])
df

Name Item Purchased  Cost
Store 1   Matthias       Dog Food  22.5
Store 1     Thomas   Kitty Litter   2.5
Store 2  Christina      Bird Seed   5.0

Indexing Operator:

In [ ]:
costs = df['Cost']
costs

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: float64

Then we can increase the cost in this series using broadcasting. 

In [ ]:
costs+=2
costs

Store 1    24.5
Store 1     4.5
Store 2     7.0
Name: Cost, dtype: float64

Now if we look at our original DataFrame, we see those costs have risen as well. This is an important consideration to watch out for. If you want to explicitly use a copy, then you should consider calling the copy method on the DataFrame for it first.

In [ ]:
df

Name Item Purchased  Cost
Store 1   Matthias       Dog Food  24.5
Store 1     Thomas   Kitty Litter   4.5
Store 2  Christina      Bird Seed   7.0

## Loading of Data

In the following, we'll be largely using smaller, moderate-sized datasets. As mentioned, a common workflow is to read the dataset in, usually from some external file.

Thankfully, Pandas has built-in support for delimited files such as CSV files as well as a variety of other data formats including relational databases, Excel, and HTML tables.

For this exercise, we'll be using a CSV file called olympics.csv, which has data from Wikipedia that contains a summary list of the medal various countries have won at the Olympics. You are able to find this file under the "Resources" folder in our GitHub. **In order for the following lines of code to work, please follow these steps:**

1. Download the file "olympics.csv" to your computer.
2. Go to Colab and click the "Files" tab on the left panel.
3. Click on "upload" and choose the olympics.csv file to be uploaded


### Reading a CSV

In [ ]:
import pandas as pd # Import pandas, if we have not already done so

Let's have a look at the content of the uploaded csv file. We can read the CSV into a DataFrame by calling the `read_csv` function of the module. When we look at the DataFrame we see that the first cell has an **NaN** in it since it's an empty value, and the rows have been automatically indexed for us.

In [ ]:
df = pd.read_csv('/content/olympics.csv')
df.head()

0         1     2     3  ...    12    13    14              15
0                NaN  № Summer  01 !  02 !  ...  01 !  02 !  03 !  Combined total
1  Afghanistan (AFG)        13     0     0  ...     0     0     2               2
2      Algeria (ALG)        12     5     2  ...     5     2     8              15
3    Argentina (ARG)        23    18    24  ...    18    24    28              70
4      Armenia (ARM)         5     1     2  ...     1     2     9              12

[5 rows x 16 columns]

It seems pretty clear that the first row of data in the DataFrame is what we really want to see as the column names. It also seems like the first column in the data is the country name, which we would like to make an index.


`read_csv` has a number of parameters that we can use to indicate to Pandas how rows and columns should be labeled.

For instance, we can use the `index_col` to indicate which column should be the index and we can also use the header parameter to indicate which row from the data file should be used as the header. 

Let's re-import that data and set the index value to be `0` which is the first column and let set a column headers to be read from the second row of data. We can do this by using the `skiprows` parameters, to tell Pandas to ignore the first row, which was made up of numeric column names.

In [ ]:
df = pd.read_csv('/content/olympics.csv', index_col = 0, skiprows=1)
df.head()

№ Summer  01 !  02 !  ...  02 !.2  03 !.2  Combined total
Afghanistan (AFG)              13     0     0  ...       0       2               2
Algeria (ALG)                  12     5     2  ...       2       8              15
Argentina (ARG)                23    18    24  ...      24      28              70
Armenia (ARM)                   5     1     2  ...       2       9              12
Australasia (ANZ) [ANZ]         2     3     4  ...       4       5              12

[5 rows x 15 columns]

Now this data came from the all time [Olympic games medal table](https://en.wikipedia.org/wiki/All-time_Olympic_Games_medal_table) on Wikipedia.

If we head to the page we could see that instead of running gold, silver and bronze in the pages, these nice little icons with a one, a two, and a three in them.

In our [raw data file](#rawdata) these were represented with the strings `01 !`, `02 !`, and so on. We see that the column values are repeated which really isn't good practice. Pandas recognize this in `01 !.1` and `01 !.2` to make things more unique.

But this labeling isn't really as clear as it could be, so we should clean up the data file. We can of course do this just by going and editing the CSV file directly, but we can also set the column names using the Pandas name property.

## Rename Columns

Panda stores a list of all of the columns in the `.columns` attribute. We can change the values of the column names by iterating over this list and calling the rename method of the data frame. 

In [ ]:
df.columns

Index(['№ Summer', '01 !', '02 !', '03 !', 'Total', '№ Winter', '01 !.1',
       '02 !.1', '03 !.1', 'Total.1', '№ Games', '01 !.2', '02 !.2', '03 !.2',
       'Combined total'],
      dtype='object')

Here we just iterate through all of the columns looking to see if they start with a `01`, `02`, `03` or numeric character. If they do, we can call rename and set the column parameters to a dictionary with the keys being the *column* we want to replace and the *value* being the new value we want.

Here we slice slice some of the old values in two, since we don't want to lose the unique appended values. We'll also set the ever-important `inplace` parameter to `True` so Pandas knows to update this data frame directly.

In [ ]:
for col in df.columns:
    if col[:2]=='01':
        df.rename(columns={col:'Gold' + col[4:]}, inplace=True)
    if col[:2]=='02':
        df.rename(columns={col:'Silver' + col[4:]}, inplace=True)
    if col[:2]=='03':
        df.rename(columns={col:'Bronze' + col[4:]}, inplace=True)
    if col[:1]=='№':
        df.rename(columns={col:'#' + col[1:]}, inplace=True) 

df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Afghanistan (AFG)              13     0       0       2      2         0   
Algeria (ALG)                  12     5       2       8     15         3   
Argentina (ARG)                23    18      24      28     70        18   
Armenia (ARM)                   5     1       2       9     12         6   
Australasia (ANZ) [ANZ]         2     3       4       5     12         0   

                         Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
Afghanistan (AFG)             0         0         0        0       13       0   
Algeria (ALG)                 0         0         0        0       15       5   
Argentina (ARG)               0         0         0        0       41      18   
Armenia (ARM)                 0         0         0        0       11       1   
Australasia (ANZ) [ANZ]       0         0         0        0        2       3   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)               0         2               2  
Algeria (ALG)                   2         8              15  
Argentina (ARG)                24        28              70  
Armenia (ARM)                   2         9              12  
Australasia (ANZ) [ANZ]         4         5              12

## Conclusion
In this section we have discussed a simple workflow to bring the data to a state where we can query it. For this we discussed the topics of *data Index*, *loading of data* and *renaming of columns*.

# Querying a DataFrame

## Boolean masking

Before we talk about how to query DataFrames, we need to talk about Boolean masking. Boolean masking is the heart of fast and efficient querying in NumPy. 

A Boolean mask is an array which can be of one dimension like a series, or two dimensions like a data frame, where each of the values in the array are either `True` or `False`. This array is essentially overlaid on top of the data structure that we're querying. And any cell aligned with the `True` value will be admitted into our final result, and any cell aligned with a `False` value will not.

Boolean masks are created by applying operators directly to the pandas Series or DataFrame objects. 

For instance, in our Olympics data set, you might be interested in seeing only those countries who have achieved a gold medal at the summer Olympics. To build a Boolean mask for this query, we project the gold column using the indexing operator and apply the **greater than operator** with a comparison value of zero. This is essentially broadcasting a comparison operator, greater than, with the results being returned as a Boolean series. 

In [ ]:
df['Gold'] > 0

Afghanistan (AFG)                               False
Algeria (ALG)                                    True
Argentina (ARG)                                  True
Armenia (ARM)                                    True
Australasia (ANZ) [ANZ]                          True
                                                ...  
Independent Olympic Participants (IOP) [IOP]    False
Zambia (ZAM) [ZAM]                              False
Zimbabwe (ZIM) [ZIM]                             True
Mixed team (ZZX) [ZZX]                           True
Totals                                           True
Name: Gold, Length: 147, dtype: bool

The resultant series is indexed where the value of each cell is either true or false depending on whether a country has won at least one gold medal, and the index is the country name.

So this builds us the Boolean mask, which is half the battle. What we want to do next is overlay that mask on the data frame.

## Where function

We can do this using the **where** function. The **where** function takes a Boolean mask as a condition, applies it to the data frame or series, and returns a new data frame or series of the same shape. Let's apply this Boolean mask to our Olympics data and create a data frame of only those countries who have won a gold at a summer games.

In [ ]:
only_gold = df.where(df['Gold'] > 0)
only_gold.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Afghanistan (AFG)             NaN   NaN     NaN     NaN    NaN       NaN   
Algeria (ALG)                12.0   5.0     2.0     8.0   15.0       3.0   
Argentina (ARG)              23.0  18.0    24.0    28.0   70.0      18.0   
Armenia (ARM)                 5.0   1.0     2.0     9.0   12.0       6.0   
Australasia (ANZ) [ANZ]       2.0   3.0     4.0     5.0   12.0       0.0   

                         Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
Afghanistan (AFG)           NaN       NaN       NaN      NaN      NaN     NaN   
Algeria (ALG)               0.0       0.0       0.0      0.0     15.0     5.0   
Argentina (ARG)             0.0       0.0       0.0      0.0     41.0    18.0   
Armenia (ARM)               0.0       0.0       0.0      0.0     11.0     1.0   
Australasia (ANZ) [ANZ]     0.0       0.0       0.0      0.0      2.0     3.0   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)             NaN       NaN             NaN  
Algeria (ALG)                 2.0       8.0            15.0  
Argentina (ARG)              24.0      28.0            70.0  
Armenia (ARM)                 2.0       9.0            12.0  
Australasia (ANZ) [ANZ]       4.0       5.0            12.0

We see that the resulting data frame keeps the original indexed values, and only data from countries that met the condition are retained. All of the countries which did not meet the condition have NaN data instead. This is okay. Most statistical functions built into the data frame object ignore values of NaN.

## Count function

For instance, if we call the df.count() on the only gold data frame:

In [ ]:
only_gold['Gold'].count()

100

we see that there are 100 countries which have had gold medals awarded at the summer games,

while if we call count on the original data frame, we see that there are 147 countries total.

In [ ]:
df['Gold'].count()

147

## Drop rows without data

Often we want to drop those rows which have no data. To do this, we can use the `dropna()` function. 

You can optionally provide dropna() the axes it should be considering. Remember that the axes is just an indicator for the columns or rows and that the default is zero, which means rows.

In [ ]:
only_gold = only_gold.dropna()
only_gold.head()
only_gold.dropna()?

# Summer   Gold  Silver  Bronze  Total  # Winter  \
Algeria (ALG)                  12.0    5.0     2.0     8.0   15.0       3.0   
Argentina (ARG)                23.0   18.0    24.0    28.0   70.0      18.0   
Armenia (ARM)                   5.0    1.0     2.0     9.0   12.0       6.0   
Australasia (ANZ) [ANZ]         2.0    3.0     4.0     5.0   12.0       0.0   
Australia (AUS) [AUS] [Z]      25.0  139.0   152.0   177.0  468.0      18.0   

                           Gold.1  Silver.1  Bronze.1  Total.1  # Games  \
Algeria (ALG)                 0.0       0.0       0.0      0.0     15.0   
Argentina (ARG)               0.0       0.0       0.0      0.0     41.0   
Armenia (ARM)                 0.0       0.0       0.0      0.0     11.0   
Australasia (ANZ) [ANZ]       0.0       0.0       0.0      0.0      2.0   
Australia (AUS) [AUS] [Z]     5.0       3.0       4.0     12.0     43.0   

                           Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                 5.0       2.0       8.0            15.0  
Argentina (ARG)              18.0      24.0      28.0            70.0  
Armenia (ARM)                 1.0       2.0       9.0            12.0  
Australasia (ANZ) [ANZ]       3.0       4.0       5.0            12.0  
Australia (AUS) [AUS] [Z]   144.0     155.0     181.0           480.0

## Indexing Operator with Boolean Mask

We do not actually have to use the where function explicitly. The pandas developers allow the indexing operator to take a Boolean mask as a value instead of just a list of column names.

Here's a more concise example of how we could query this data frame. You'll notice that there are no NaNs when you query the data frame in this manner. pandas automatically filters out the rows with now values.

In [ ]:
only_gold = df.loc[df['Gold'] > 0]
only_gold.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Algeria (ALG)                    12     5       2       8     15         3   
Argentina (ARG)                  23    18      24      28     70        18   
Armenia (ARM)                     5     1       2       9     12         6   
Australasia (ANZ) [ANZ]           2     3       4       5     12         0   
Australia (AUS) [AUS] [Z]        25   139     152     177    468        18   

                           Gold.1  Silver.1  Bronze.1  Total.1  # Games  \
Algeria (ALG)                   0         0         0        0       15   
Argentina (ARG)                 0         0         0        0       41   
Armenia (ARM)                   0         0         0        0       11   
Australasia (ANZ) [ANZ]         0         0         0        0        2   
Australia (AUS) [AUS] [Z]       5         3         4       12       43   

                           Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                   5         2         8              15  
Argentina (ARG)                18        24        28              70  
Armenia (ARM)                   1         2         9              12  
Australasia (ANZ) [ANZ]         3         4         5              12  
Australia (AUS) [AUS] [Z]     144       155       181             480

The syntax might look a little messy, especially if you're not used to programming languages with overloaded operators, but the result is that you're able to filter and reduce data frames relatively quickly.

## Chain Statements

One more thing to keep in mind if you're not used to Boolean or bit masking for data reduction. The output of two Boolean masks being compared with logical operators is another Boolean mask. This means that you can chain together a bunch of and/or statements in order to create more complex queries, and the result is a single Boolean mask.

For instance, we could create a mask for all of those countries who have received a gold in the summer Olympics and logically order that with all of those countries who have received a gold in the winter Olympics. If we apply this to the data frame and use the length function to see how many rows there are, we see that there are 101 countries which have won a gold metal at some time.

In [ ]:
len(df[(df['Gold'] > 0) | (df['Gold.1'] > 0)])

101

Another example. Have there been any countries who have only won a gold in the winter Olympics and never in the summer Olympics? 

Here's one way to answer that.

In [ ]:
df[(df['Gold.1'] > 0) & (df['Gold'] == 0)]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Liechtenstein (LIE)        16     0       0       0      0        18       2   

                     Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  \
Liechtenstein (LIE)         2         5        9       34       2         2   

                     Bronze.2  Combined total  
Liechtenstein (LIE)         5               9

Our neighbouring country Liechtenstein

Extremely important, and often an issue for new users, is to remember that each Boolean mask needs to be encased in parenthesis because of the order of operations. This can cause no end of frustration if you're not used to it, so be careful.


## Conclusion

In this lecture, we took a look at boolean masking. In particular, we learned about the where function, count function, droping rows and indexing operator with boolean masking.

# Indexing DataFrames

As we have  seen, both Series and DataFrames can have indices applied to them. The index is essentially a row level label, and we know that rows correspond to axis zero. In our Olympics data, we indexed the data frame by the name of the country. Indices can either be inferred, such as when we create a new series without an index, in which case we get numeric values, or they can be set explicitly, like when we use the dictionary object to create the series, or when we loaded data from the CSV file and specified the header. 

In [ ]:
df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Afghanistan (AFG)              13     0       0       2      2         0   
Algeria (ALG)                  12     5       2       8     15         3   
Argentina (ARG)                23    18      24      28     70        18   
Armenia (ARM)                   5     1       2       9     12         6   
Australasia (ANZ) [ANZ]         2     3       4       5     12         0   

                         Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
Afghanistan (AFG)             0         0         0        0       13       0   
Algeria (ALG)                 0         0         0        0       15       5   
Argentina (ARG)               0         0         0        0       41      18   
Armenia (ARM)                 0         0         0        0       11       1   
Australasia (ANZ) [ANZ]       0         0         0        0        2       3   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)               0         2               2  
Algeria (ALG)                   2         8              15  
Argentina (ARG)                24        28              70  
Armenia (ARM)                   2         9              12  
Australasia (ANZ) [ANZ]         4         5              12

## Set an Index

Another option for setting an index is to use the `set_index()` function. This function takes a list of columns and promotes those columns to an index. `set_index()` is a **destructive process**, it doesn't keep the current index. If you want to keep the current index, you need to manually create a new column and copy into it values from the index attribute. 

Let's go back to our Olympics DataFrame. Let's say that we don't want to index the DataFrame by countries, but instead want to index by the number of gold medals that were won at summer games. First we need to preserve the country information into a new column. We can do this using the indexing operator or the string that has the column label. Then we can use  `set_index()` to set the index of the column to summer gold medal wins.

In [ ]:
df['country'] = df.index
df = df.set_index('Gold')
df.head()

# Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  Bronze.1  \
Gold                                                                          
0           13       0       2      2         0       0         0         0   
5           12       2       8     15         3       0         0         0   
18          23      24      28     70        18       0         0         0   
1            5       2       9     12         6       0         0         0   
3            2       4       5     12         0       0         0         0   

      Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total  \
Gold                                                                 
0           0       13       0         0         2               2   
5           0       15       5         2         8              15   
18          0       41      18        24        28              70   
1           0       11       1         2         9              12   
3           0        2       3         4         5              12   

                      country  
Gold                           
0           Afghanistan (AFG)  
5               Algeria (ALG)  
18            Argentina (ARG)  
1               Armenia (ARM)  
3     Australasia (ANZ) [ANZ]

You'll see that when we create a new index from an existing column it appears that a new first row has been added with empty values. This isn't quite what's happening. And we know this in part because an empty value is actually rendered either as a none or an `NaN` if the data type of the column is numeric. What's actually happened is that the index has a name. Whatever the column name was in the Jupiter notebook has just provided this in the output.

## Reset Index

We can get rid of the index completely by calling the function `reset_index()`. This promotes the index into a column and creates a default numbered index.

In [ ]:
df = df.reset_index()
df.head()

Gold  # Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  \
0     0        13       0       2      2         0       0         0   
1     5        12       2       8     15         3       0         0   
2    18        23      24      28     70        18       0         0   
3     1         5       2       9     12         6       0         0   
4     3         2       4       5     12         0       0         0   

   Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total  \
0         0        0       13       0         0         2               2   
1         0        0       15       5         2         8              15   
2         0        0       41      18        24        28              70   
3         0        0       11       1         2         9              12   
4         0        0        2       3         4         5              12   

                   country  
0        Afghanistan (AFG)  
1            Algeria (ALG)  
2          Argentina (ARG)  
3            Armenia (ARM)  
4  Australasia (ANZ) [ANZ]

# Sorting and Min Max

Data exploration is an important aspect of Data Science. Before we decide do Data Analysis or Maschine Learnig, we must have an idea of what our data contains. The Pandas library is equipped with a number of useful functions for this very purpose.

## Counting Values

The `value_counts()` method returns a Series containing the counts of unique values. This means, for any column in a dataframe, this method returns the count of unique entries in that column.

The following example uses `value_counts()` to determine which country appears most often.

In [ ]:
import pandas as pd

# Make example dataframe
df = pd.DataFrame([(1, 'Germany'),
                   (2, 'France'),
                   (3, 'Indonesia'),
                   (4, 'France'),
                   (5, 'France'),
                   (6, 'Germany'),
                   (7, 'UK'),
                   ],
                  columns=['groupid', 'country'],
                  index=['a', 'b', 'c', 'd', 'e', 'f', 'g'])

df

groupid    country
a        1    Germany
b        2     France
c        3  Indonesia
d        4     France
e        5     France
f        6    Germany
g        7         UK

In [ ]:
# What you're looking for
values = df['country'].value_counts().keys().tolist()
counts = df['country'].value_counts().tolist()
print(df['country'].value_counts())
print("\n")
print(values)
print(counts)

France       3
Germany      2
Indonesia    1
UK           1
Name: country, dtype: int64


['France', 'Germany', 'Indonesia', 'UK']
[3, 2, 1, 1]


'France' is the country that appears most frequently with a frequency of 3 in the DataFrame.

Panda's `keys()` function returns the *info axis* for the Panda object. If the Pandas object is a Series, it returns the index. In the given example, applied to value_counts(), it returns the corresponding values for the given counts.

## Sorting Index

Sorting can be done in Pandas by using the `sort_index()` or `sort_values()` function.

For this exercise we will be using another csv file that contains a list of movies from the International Movie Database (IMDb). This file can, once again, be found within the Resource folder on GitHub and must be uploaded to Google Colab.

Pandas `sort_index()` function sorts objects by labels along the given axis.
Basically the sorting algorithm is applied on the axis labels rather than the actual data in the dataframe and based on that the data is rearranged.

In [ ]:
import pandas as pd

df_imdb = pd.read_csv('/content/IMDBdata_hotlist.csv')
df_imdb.head()

Movie_Title  YR_Released  ...              Record  Runtime
0  The Shawshank Redemption         1994  ...  Top 250 Eng Movies    142.0
1             The Godfather         1972  ...  Top 250 Eng Movies    175.0
2    The Godfather: Part II         1974  ...  Top 250 Eng Movies    202.0
3           The Dark Knight         2008  ...  Top 250 Eng Movies    152.0
4              12 Angry Men         1957  ...  Top 250 Eng Movies     96.0

[5 rows x 7 columns]

What are the three best reviewed movies in the database? To obtain a ranking of the best, you must be able to sort. Here we sort on the dataframe directly (`inplace=True`), and we want to see the highest ranked films first, so we proceed in descending order, i.e. `ascending=False`.

In [ ]:
cdf_imdb = df_imdb.copy()
cdf_imdb = cdf_imdb.set_index('Rating')
cdf_imdb.sort_index(inplace=True, ascending=False)
cdf_imdb.head()

Movie_Title  YR_Released  ...              Record Runtime
Rating                                         ...                            
9.2     The Shawshank Redemption         1994  ...  Top 250 Eng Movies   142.0
9.2                The Godfather         1972  ...  Top 250 Eng Movies   175.0
9.0       The Godfather: Part II         1974  ...  Top 250 Eng Movies   202.0
9.0              The Dark Knight         2008  ...  Top 250 Eng Movies   152.0
8.9                 12 Angry Men         1957  ...  Top 250 Eng Movies    96.0

[5 rows x 6 columns]

Here we show two possibilities how the first three entries can be presented. Either with the `head()` method or with the `iloc` operator that uses slicing.

In [ ]:
# Show the three best movies
cdf_imdb.head(3)

Movie_Title  YR_Released  ...              Record Runtime
Rating                                         ...                            
9.2     The Shawshank Redemption         1994  ...  Top 250 Eng Movies   142.0
9.2                The Godfather         1972  ...  Top 250 Eng Movies   175.0
9.0       The Godfather: Part II         1974  ...  Top 250 Eng Movies   202.0

[3 rows x 6 columns]

In [ ]:
# Show the three best movies
cdf_imdb.iloc[:3]

Movie_Title  YR_Released  ...              Record Runtime
Rating                                         ...                            
9.2     The Shawshank Redemption         1994  ...  Top 250 Eng Movies   142.0
9.2                The Godfather         1972  ...  Top 250 Eng Movies   175.0
9.0       The Godfather: Part II         1974  ...  Top 250 Eng Movies   202.0

[3 rows x 6 columns]

The best three movies in the database are `The Shawshank Redemption` and `The Godfather` I \& II

## Sorting

In the previous section we indexed the film list according to the ranking and sorted the ranking list correspondingly.

Here we use a very similar function `sort_values()` that lets you sort a table by any column. The `by` parameter is the name or list of names to sort by. If no parameter is given (i.e. axis is 0) then by contain column labels.


In [ ]:
cdf_imdb = df_imdb.copy()
cdf_imdb = cdf_imdb.sort_values(by='Rating', ascending=False)
cdf_imdb

Movie_Title  YR_Released  ...              Record  Runtime
0        The Shawshank Redemption         1994  ...  Top 250 Eng Movies    142.0
1                   The Godfather         1972  ...  Top 250 Eng Movies    175.0
2          The Godfather: Part II         1974  ...  Top 250 Eng Movies    202.0
3                 The Dark Knight         2008  ...  Top 250 Eng Movies    152.0
4                    12 Angry Men         1957  ...  Top 250 Eng Movies     96.0
..                            ...          ...  ...                 ...      ...
832                  Pledge This!         2006  ...   Bottom 250 Movies     91.0
833  Superbabies: Baby Geniuses 2         2004  ...   Bottom 250 Movies     88.0
834             Daniel the Wizard         2004  ...   Bottom 250 Movies     81.0
835              Saving Christmas         2014  ...   Bottom 250 Movies     80.0
836             Code Name: K.O.Z.         2015  ...   Bottom 250 Movies    114.0

[837 rows x 7 columns]

## Min Max

Pandas has  min and max functions, used to find the minimum value and maximum value of any given DataFrame.

In [ ]:
df = pd.read_csv("~/data/ds4/subject_grades.csv")

In [ ]:
df.head(4)

Physics  Biology  Maths
0       80       78     70
1       60       98     45
2       34       56     65
3       56       65     32

Show descriptive statistics of all columns

In [ ]:
df.describe()

Physics    Biology      Maths
count  20.000000  20.000000  20.000000
mean   66.450000  71.200000  69.500000
std    22.705147  19.813871  17.276391
min    12.000000  23.000000  32.000000
25%    53.250000  56.000000  60.000000
50%    72.000000  71.500000  68.000000
75%    80.000000  82.000000  81.000000
max    98.000000  98.000000  99.000000

Show descriptive statistics only for the column of “Maths”

df['Maths'].describe()

In [ ]:
maths_mean = df['Maths'].mean()
physics_max = df['Physics'].max()
biology_min = df['Biology'].min()
print(maths_mean)
print(physics_max)
print(biology_min)

69.5
98
23


Calculate Min und Max per Row

In [ ]:
df['Min'] = df.min(axis=1)
df['Max'] = df.max(axis=1)

In [ ]:
df.head(5)

Physics  Biology  Maths  Min  Max
0       80       78     70   70   80
1       60       98     45   45   98
2       34       56     65   34   65
3       56       65     32   32   65
4       98       23     55   23   98

Calculate Min und Max only for Physics and Maths Row

In [ ]:
df['PMMin'] = df.loc[:, ['Physics','Maths']].min(axis=1)
df['PMMax'] = df.loc[:, ['Physics','Maths']].max(axis=1)

In [ ]:
df.head(5)

Physics  Biology  Maths  Min  Max  PMMin  PMMax
0       80       78     70   70   80     70     80
1       60       98     45   45   98     45     60
2       34       56     65   34   65     34     65
3       56       65     32   32   65     32     56
4       98       23     55   23   98     55     98

## idxmin() and idxmax()

In a nutshell, those functions will return the ID (the index position) of the smallest `idxmin()` and largest `idxmax()` item. 

For example, to return to the already mentioned reviewed movie database. Let's assume that we only want to find one of the best and one of the worst rated movies, then we can do the following:

In [ ]:
import pandas as pd
cdf_imdb = pd.read_csv('/content/IMDBdata_hotlist.csv')

In [ ]:
bestindex = cdf_imdb['Rating'].idxmax()

# One of the best movies, i.e. highest rating
bestmovie = cdf_imdb.iloc[bestindex]['Movie_Title']

print("One of the best movies is «{}»".format(bestmovie))

One of the best movies is «The Shawshank Redemption»


Here we apply the function `idxmax()` to the `Rating` column and get the corresponding index of one of the movies with the maximum rating. This index can be used with the `iloc` operator to get the data frame row of the movie. Using a operation chaining, we select the movie title column to get the movie name as a string.

In [ ]:
print("One of the worst movies is «{}»".format(cdf_imdb.iloc[cdf_imdb['Rating'].idxmin()]['Movie_Title']))

One of the worst movies is «Code Name: K.O.Z.»


Here we use a comprehensive code, essentially using `idxmin` in the style of the previous `idxmax` example. This means that we use `idxmin()`` to get the index of one of the movies with the lowest rating.

You can also find the function `argmax()` in source code on the Web or elsewhere. But there is not much difference between `idxmax()` and `argmax()`. idxmax returns the index name of the maximum value. argmax does the same. However idxmax is a Pandas function, so it can take Series and DataFrame. argmax is a Numpy function.

